File: MST.java
//Implement new comparator to prioritize nodes by cost in pq
static class CompareNodes implements Comparator<Node> {
    public int compare(Node n1, Node n2) {
        if (n1.cost > n2.cost) {
            return 1;
        } else if (n1.cost < n2.cost) {
            return -1;
        } else {
            return 0;
        }
    }
}

// Node class for for the grid of nodes in Path.
// Contains coordinates, and cost to get to that node, and a reference to the parent node.
public class Node{
    private int n;
    private int x;
    private int y;
    private int cost;
    private Node parent;
    private boolean visited;
/* --------------------------------- Methods -------------------------------- */
    // Constructor
    public Node(int n, int x, int y, int cost, Node parent){
        this.n = n;
        this.x = x;
        this.y = y;
        this.cost = cost;
        this.parent = parent;
    }
    // Getters
    public int getN(){
        return n;
    }
    public int getX(){
        return x;
    }
    public int getY(){
        return y;
    }
    public int getCost(){
        return cost;
    }
    public Node getParent(){
        return parent;
    }
    public boolean getVisited(){
        return visited;
    }
    // Setters
    public void setCost(int cost){
        this.cost = cost;
    }
    public void setParent(Node parent){
        this.parent = parent;
    }
    public void setVisited(boolean visited){
        this.visited = visited;
    }

}

//Path class for MST. Has a size, and adjacency matrix, a vertex count.
public class Path {
    private int size;
    private int[][] adjacencyMatrix;
    private int nodeCount;
    private Node[] nodes;

    //Constructor for Path.
    public Path(int size){
        this.size = size;
        this.nodeCount = size*size;
        this.nodes = new Node[nodeCount];
        for(int i = 0; i < nodeCount; i++){
            this.nodes[i] = new Node(i, i%this.size, i/this.size, Integer.MAX_VALUE, null);
        }
        this.adjacencyMatrix = new int[size*size][size*size];
        //initialize adjacency matrix to -1
        for(int i = 0; i < size; i++){
            for(int j = 0; j < size; j++){
                adjacencyMatrix[i][j] = -1;
            }
        }
        this.connect_adjacent_nodes_in_grid();
        this.MST_by_prims();
    }
    private void addEdge(int i, int j, int weight){
        adjacencyMatrix[i][j] = weight;
    }
    private void connect_adjacent_nodes_in_grid(){
        for(int node=0; i<nodeCount; i++){
            int x = i%size;
            int y = i/size;
        }
        if(x>0){
            this.addEdge(i,i-1, (int)(Math.random()*50));
        }
        if(x<size){
            this.addEdge(i,i+1, (int)(Math.random()*50));
        }
        if(y>0){
            this.addEdge(i, i-size, (int)(Math.random()*50));
        }
        if(y<size){
            this.addEdge(i, i+size, (int)(Math.random()*50));
        }
        
    }
    private void MST_by_prims(){
        // create priority queue to store nodeCount nodes
        PriorityQueue<Node> pq = new PriorityQueue<Node>(new CompareNodes());
        // add nodes to pq
        for(int i = 0; i < nodeCount; i++){
            pq.add(nodes[i]);
        }
        //set the first node to have a cost of 0
        pq.peek().setCost(0);
        //while the queue is not empty
        while(!pq.isEmpty()){
            //get the node with the lowest cost
            pq.peek().setVisited(true);
            //pop the node with the lowest cost
            Node u = pq.poll();
            //get neighbors of u
            for(int i = 0; i < nodeCount; i++){
                //Todo: change to using arithmetic instead of for/if
                if(adjacencyMatrix[u.getN()][i] != -1){
                    //if the neighbor is not visited
                    if(!pq.peek().getVisited()){
                        //if the weight between u and the neighbor is less than the neighbor's known cost
                        if(adjacencyMatrix[u.getN()][i] < pq.peek().getCost()){
                            //set the neighbor's parent to u
                            pq.peek().setParent(u);
                            //set the neighbor's new cost
                            pq.peek().setCost(adjacencyMatrix[u.getN()][i]);
                        }
                    }
                }

            }
        }




        }
    //getter for nodes
    public Node[] getNodes(){
        return nodes;
    }
}
//main class
class Main{
public static void main(String[] args){
    Path p = new Path(3);
    Node[] nodes = p.getNodes();
    for(int i = 0; i < nodes.length; i++){
        System.out.println(nodes[i].getN() + " " + nodes[i].getParent());
    }
}
}

File: MST.py
import matplotlib.pyplot as plt
from math import inf
from random import randint
import argparse

#constants
maze_len=10

class Path():
    def __init__(self, size):
        self.size = size
        self.adjacency_matrix = [[-1 for x in range(size**2)] for y in range(size**2)]
        self.vertc = size * size
        self.connect_adjacent_nodes_in_grid()
        self.P = self.MST_by_prims()
        self.directions = self.parent_array_to_directions()
        
    def add_edge(self, v1, v2, weight):
        self.adjacency_matrix[v1][v2] = weight
        self.adjacency_matrix[v2][v1] = weight
    
    def connect_adjacent_nodes_in_grid(self):
        for vertex in range(self.vertc):
            x = vertex % self.size
            y = vertex // self.size
            # connect to left if not on left edge
            if x > 0:
                self.add_edge(vertex, vertex-1, randint(1,50))
            # connect to right if not on right edge
            if x < self.size - 1:
                self.add_edge(vertex, vertex+1, randint(1,50))
            # connect to top if not on top edge
            if y > 0:
                self.add_edge(vertex, vertex-self.size, randint(1,50))
            # connect to bottom if not on bottom edge
            if y < self.size - 1:
                self.add_edge(vertex, vertex+self.size, randint(1,50))
    def plot_mst(self):
        # draw grid of points
        x = [i % self.size for i in range(self.vertc)]
        y = [i // self.size for i in range(self.vertc)]
        plt.scatter(x, y, s=15)
        # draw edges
        for i in range(self.vertc):
            # Draw edge from i to P[i]
            if self.P[i] is not None:
                plt.plot([i % self.size, self.P[i] % self.size], [i // self.size, self.P[i] // self.size], color='r')
            
        plt.show()
        
    def parent_array_to_directions(self):
        directions = []
        for i in range(self.vertc):
            if self.P[i] is not None:
                x = i % self.size
                y = i // self.size
                p_x = self.P[i] % self.size
                p_y = self.P[i] // self.size
                if x == p_x:
                    if y > p_y:
                        directions.append('U')
                        print(f'{self.P[i]} moves up to {i}')
                    else:
                        directions.append('D')
                        print(f'{self.P[i]} moves down to {i}')
                elif y == p_y:
                    if x < p_x:
                        directions.append('L')
                        print(f'{self.P[i]} moves left to {i}')
                    else:
                        directions.append('R')
                        print(f'{self.P[i]} moves right to {i}')
            else:
                directions.append('n')
                print(f'{self.P[i]} does not move to {i}')
                        
        return directions
        
    def plot_mst_animated(self):
        x = [i % self.size for i in range(self.vertc)]
        y = [i // self.size for i in range(self.vertc)]
        plt.scatter(x, y, s=15)
        for i in range(len(self.P)):
            if self.P[i] is not None:
                plt.plot([i % self.size, self.P[i] % self.size], [i // self.size, self.P[i] // self.size], color='r')
                # write direction direct[i] at point
                #plt.text((i % self.size) + 0.5, (i // self.size) + 0.5, direct[i], fontsize=10)
                print(f'{self.P[i]} moves to {i}')
                    
                plt.pause(0.1)
        plt.show()
        
        
    def plot_by_directions(self):
        x = [i % self.size for i in range(self.vertc)]
        y = [i // self.size for i in range(self.vertc)]
        plt.scatter(x, y, s=15)
        # label each point with its id
        for i in range(self.vertc):
            plt.text((i % self.size) + 0.2, (i // self.size) + 0.2, str(i), fontsize=10)
        for i in range(len(self.P)):
            if self.P[i] is not None:
                if self.directions[i] == 'D':
                    print(f'{self.P[i]} moves down')
                    plt.plot([self.P[i] % self.size, self.P[i] % self.size], [self.P[i] // self.size,self.P[i] // self.size - 1], color='r')
                elif self.directions[i] == 'R':
                    print(f'{self.P[i]} moves right')
                    plt.plot([self.P[i] % self.size, self.P[i] % self.size + 1], [self.P[i] // self.size, self.P[i] // self.size], color='r')
                elif self.directions[i] == 'U':
                    print(f'{self.P[i]} moves up')
                    plt.plot([self.P[i] % self.size, self.P[i] % self.size], [self.P[i] // self.size, self.P[i] // self.size + 1], color='r')
                elif self.directions[i] == 'L':
                    print(f'{self.P[i]} moves left')
                    plt.plot([self.P[i] % self.size, self.P[i] % self.size - 1], [self.P[i] // self.size, self.P[i] // self.size], color='r')
                
            else:
                print('at start')
        plt.show()
            
    def MST_by_prims(self):
        # each node is a vertex, and has ID (size*y + x)
        visited = set()
        P = [None for x in range(self.vertc)]
        # keys represent the vertices, and values represent the known cost of reaching that vertex
        keys = dict.fromkeys(range(self.vertc), inf)
        keys[0] = 0
        # alternatively, while(keys is not empty)
        while (len(visited) < self.vertc):
            # find index of cheapest key
            u = min(keys, key=keys.get)
            # add u to visited nodes
            visited.add(u)
            # remove u from keys
            del keys[u]
            # get neighbors of u
            u_x = u % self.size
            u_y = u // self.size
            neighbors = []
            if u_x > 0:
                neighbors.append(u - 1)
            if u_x < self.size - 1:
                neighbors.append(u + 1)
            if u_y > 0:
                neighbors.append(u - self.size)
            if u_y < self.size - 1:
                neighbors.append(u + self.size)
            # for each neighbor of u
            for v in neighbors:
                # if v is not visited
                if v not in visited:
                    # get weight of edge u-v
                    u_v = self.adjacency_matrix[u][v]
                    # if weight of edge u-v is less than the known cost of reaching v
                    if u_v < keys[v]:
                        # update the known cost of reaching v
                        keys[v] = u_v
                        # update the parent of v so the path to v is through u
                        P[v] = u
        # return the path
        return P




# Main
if __name__ == '__main__':
    # Parse arguments
    parser = argparse.ArgumentParser(description='Generates a random grid-spanning tree.')
    # get the size of the maze
    parser.add_argument('-s', '--size', type=int, default=10, help='Size of the maze.')
    args = parser.parse_args()
    #maze length
    size = args.size
    maze = Path(size)
    maze.plot_mst_animated()
    #maze.plot_mst()